covenants/src/lib.rs

```diff
@@ -0,0 +1,1160 @@
+        // lender agrees with the repayment condition
+
+        let collateral_script = loan_contract(
+            self.keypair.1,
+            loan_response.lender_pk,
+            loan_response.timelock,
+            loan_response.repayment_principal_output.clone(),
+        )?;
+
+        dbg!(&collateral_script);
+
+        let collateral_address = Address::p2wsh(&collateral_script, None, &AddressParams::ELEMENTS);
+        let collateral_script_pubkey = collateral_address.script_pubkey();
+        let collateral_blinding_sk = loan_response.repayment_collateral_input.blinding_key;
+        transaction
+            .output
+            .iter()
+            .find_map(|out| match out.to_confidential() {
+                Some(conf) => {
+                    let unblinded_out = conf.unblind(secp, collateral_blinding_sk).ok()?;
+                    let predicate = unblinded_out.asset == self.bitcoin_asset_id
+                        && unblinded_out.value == self.collateral_amount.as_sat()
+                        && out.script_pubkey == collateral_script_pubkey;
+
+                    predicate.then(|| out)
+                }
+                None => None,
+            })
+            .context("no collateral txout")?;
+
+        let collateral_input_amount = self
+            .collateral_inputs
+            .iter()
+            .map(|input| input.clone().into_unblinded_input(secp))
+            .try_fold(0, |sum, input| {
+                input.map(|input| sum + input.unblinded.value).ok()
+            })
+            .context("could not sum collateral inputs")?;
+        let tx_fee = Amount::from_sat(
+            estimate_virtual_size(transaction.input.len() as u64, 4)
+                * self.fee_sats_per_vbyte.as_sat(),
+        );
+        let collateral_change_amount = Amount::from_sat(collateral_input_amount)
+            .checked_sub(self.collateral_amount)
+            .map(|a| a.checked_sub(tx_fee))
+            .flatten()
+            .with_context(|| {
+                format!(
+                    "cannot pay for output {} and fee {} with input {}",
+                    self.collateral_amount, tx_fee, collateral_input_amount,
+                )
+            })?;
+
+        transaction
+            .output
+            .iter()
+            .find_map(|out| match out.to_confidential() {
+                Some(conf) => {
+                    let unblinded_out = conf.unblind(secp, self.address_blinding_sk).ok()?;
+                    let predicate = unblinded_out.asset == self.bitcoin_asset_id
+                        && unblinded_out.value == collateral_change_amount.as_sat()
+                        && out.script_pubkey == self.address.script_pubkey();
+
+                    predicate.then(|| out)
+                }
+                None => None,
+            })
+            .context("no collateral change txout")?;
+
+        Ok(Borrower1 {
+            keypair: self.keypair,
```

Comment:
nit: renaming to what this key is used.

Reply:
What did you have in mind? Isn't this just the borrower's keypair?

Reply:
Yes, but for what? I honestly forgot for what atm :D